File: src/syntax/lexer.ts
import { Token, TType, tokens, keywords, symbols } from './token'

const SYMBOLS = '.%^&|:=~+-*<>!/'

interface Lexer {
  source: string
  line: number
  prev: number
  curr: number
  mode: Mode,
  tokens: Token[]
}

type Mode =
  'regular'
| 'string'
| 'template'

function lexer(source: string): Lexer {
  return {
    source,
    line: 1,
    curr: 0,
    prev: 0,
    mode: 'regular',
    tokens: [],
  }
}

function token(lexer: Lexer, type: TType, value?: string): Lexer {
  lexer.tokens.push({
    type,
    value,
    line: lexer.line,
    span: {
      start: lexer.prev,
      offset: lexer.curr - lexer.prev,
    }
  })
  return lexer
}

function bump(lexer: Lexer): Lexer {
  lexer.curr += 1
  return lexer
}

function line(lexer: Lexer): Lexer {
  lexer.line += 1
  return bump(lexer)
}

function char(lexer: Lexer) {
  return lexer.source[lexer.curr]
}

function peek(lexer: Lexer) {
  return lexer.source[lexer.curr + 1]
}

function slice(lexer: Lexer) {
  return lexer.source.slice(lexer.prev, lexer.curr)
}

function done(lexer: Lexer) {
  return lexer.curr >= lexer.source.length
}

function numeric(ch: string) {
  return ch >= '0' && ch <= '9'
}

function lower(ch: string) {
  return ch >= 'a' && ch <= 'z' || ch === '_'
}

function upper(ch: string) {
  return ch >= 'A' && ch <= 'Z'
}

function alpha(ch: string) {
  return lower(ch) || upper(ch)
}

function alphanumeric(ch: string) {
  return lower(ch) || upper(ch) || numeric(ch)
}

function symbolic(ch: string) {
  return SYMBOLS.includes(ch)
}

function space(lexer: Lexer): Lexer {
  let ch = char(lexer)
  while (ch === ' '
      || ch === '\r'
      || ch === '\t') {
    ch = char(bump(lexer))
  }
  return next(lexer)
}

function comment(lexer: Lexer): Lexer {
  while (char(lexer) !== '\n' && !done(lexer)) {
    bump(lexer)
  }
  return lexer
}

function number(lexer: Lexer, prefix = false): Lexer {
  if (prefix) {
    bump(lexer)
  }
  let float = false
  while (numeric(char(lexer))) {
    bump(lexer)
    if (char(lexer) === '.' && !float) {
      float = true
      bump(lexer)
    }
  }
  return token(lexer, tokens.number, slice(lexer))
}

function identifier(lexer: Lexer): Lexer {

  while (alphanumeric(char(lexer))) {
    bump(lexer)
  }

  switch (char(lexer)) {
    case '?':
    case '!': {
      bump(lexer)
    }
    break
    case '\'': {
      while (char(lexer) === '\'') {
        bump(lexer)
      }
    }
    break
  }

  const value = slice(lexer)

  const keyword = keywords[value]

  if (keyword) {
    return token(lexer, keyword)
  } else {
    return token(lexer, upper(value[0]) ? tokens.upper : tokens.lower, value)
  }
}

function wildcard(lexer: Lexer): Lexer {
  bump(lexer)
  return token(lexer, tokens.underscore)
}

function operator(lexer: Lexer): Lexer {
  while (symbolic(char(lexer))) {
    bump(lexer)
  }

  const operator = symbols[slice(lexer)]

  if (operator != undefined) {
    return token(lexer, operator)
  } else {
    return error(lexer, `Invalid operator '${slice(lexer)}'`)
  }
}

// TODO: collect the errors and handle lexer/parser recovery
function error(lexer: Lexer, msg: string): never {
  throw new Error(`${msg} at ${lexer.line}`)
}

function next(lexer: Lexer): Lexer {
  lexer.prev = lexer.curr

  if (lexer.mode === 'string') {
    return template(lexer, false)
  }

  const ch = char(lexer)

  switch (ch) {
    case ' ':
    case '\t':
    case '\r': 
      return space(lexer)
    case '\n':
      return line(lexer)
    case ',':
      return token(bump(lexer), tokens.comma)
    case ';':
      return token(bump(lexer), tokens.semicolon)
    case '(':
      return token(bump(lexer), tokens.lparen)
    case ')':
      return token(bump(lexer), tokens.rparen)
    case '[':
      return token(bump(lexer), tokens.lbracket)
    case ']':
      return token(bump(lexer), tokens.rbracket)
    case '{':
      return token(bump(lexer), tokens.lbrace)
    case '}': {
      if (lexer.mode === 'template') {
          lexer.mode = 'string'
      }
      return token(bump(lexer), tokens.rbrace)
    }
    case '"':
      return template(lexer, true)
    case '+':
    case '-':
      if (numeric(peek(lexer))) {
        return number(lexer, true)
      } else {
        return operator(lexer)
      }
    case '/':
      if (peek(lexer) === '/') {
        return comment(lexer)
      } else {
        return operator(lexer)
      }
    case '_':
      if (alpha(peek(lexer))) {
        return identifier(lexer)
      } else {
        return wildcard(lexer)
      }
    default: break
  }
  if (alpha(ch)) {
    return identifier(lexer)
  } else if (numeric(ch)) {
    return number(lexer)
  } else if (symbolic(ch)) {
    return operator(lexer)
  } else {
    bump(lexer)
    return error(lexer, `Invalid character '${slice(lexer)}'`)
  }
}

function template(lexer: Lexer, start: boolean): Lexer {
  lexer.prev = lexer.curr

  while (!done(lexer)) {
    switch (char(lexer)) {
      case '"': {
        if (lexer.mode === 'regular') {
          bump(lexer)
          lexer.mode = 'string'
        } else {
          lexer.mode = 'regular'
          bump(lexer)
          const kind = start ? tokens.string : tokens.stringFinish
          return token(lexer, kind, slice(lexer))
        }
      }
        break
      case '{': {
        if (peek(lexer) === '{') {
          bump(lexer)
          bump(lexer)
        } else {
          const kind = start ? tokens.stringStart : tokens.stringPart
          lexer.mode = 'template'
          return token(lexer, kind, slice(lexer))
        }
      }
        break
      case '\n':
        return error(lexer, 'Unterminated string')
      case '\\': {
        bump(lexer)
        switch (char(lexer)) {
          case 'n': case 'r':
          case 't': case 'v':
          case 'a': case 'b':
          case '"': case '0':
          case '\\': {
            bump(lexer)
          }
            break
          case 'u':
            return error(lexer, 'TODO Validate unicode escape')
          case 'x':
            return error(lexer, 'TODO Validate binary escape')
          default:
            return error(lexer, `Invalid escape character ${char(lexer)}`)
        }
      }
        break
      default:
        bump(lexer)
    }
  }

  return lexer
}

export function tokenize(source: string) {
  const tokenizer = lexer(source)

  while (!done(tokenizer)) { next(tokenizer) }

  return tokenizer.tokens
}

File: src/syntax/lexer.rs
use crate::source::{Source, Span};

use super::token::{Token, TokenKind};

#[derive(Debug, Clone)]
pub struct Lexer<'l> {
    source: &'l Source,
    bytes: &'l [u8],
    curr: u8,
    peek: u8,
    index: usize,
    start: usize,
}

impl<'l> Lexer<'l> {
    fn new(source: &'l Source) -> Lexer<'l> {
        let mut lexer = Lexer {
            source,
            bytes: source.content.as_bytes(),
            curr: 0,
            peek: 0,
            index: 0,
            start: 0,
        };
        lexer.curr = lexer.at(0);
        lexer.peek = lexer.at(1);
        lexer
    }

    fn align(&mut self) {
        self.start = self.index;
    }

    fn bump(&mut self) -> u8 {
        self.index += 1;
        self.curr = self.peek;
        self.peek = self.at(self.index + 1);
        self.curr
    }

    fn at(&self, idx: usize) -> u8 {
        if idx < self.bytes.len() {
            self.bytes[idx]
        } else {
            b'\0'
        }
    }

    fn slice(&self) -> &str {
        &self.source.content[self.start..self.index]
    }

    fn span(&self) -> Span {
        Span::new(self.start as u32, self.index as u32)
    }

    fn bump_while<P>(&mut self, pred: P)
    where
        P: Fn(u8) -> bool,
    {
        while pred(self.curr) {
            self.bump();
        }
    }

    fn kind(&mut self) -> TokenKind {
        self.trivia();
        self.align();

        let kind = match self.curr {
            b'a'..=b'z' | b'A'..=b'Z' | b'_' => {
                return self.ident();
            }
            b'0'..=b'9' => {
                return self.number();
            }
            b'"' => {
                return self.string();
            }
            b'\n' => {
                return self.newline();
            }
            _ if self.is_symbol(self.curr) => {
                return self.operator();
            }
            b'(' => TokenKind::LParen,
            b')' => TokenKind::RParen,
            b'{' => TokenKind::LBrace,
            b'}' => TokenKind::RBrace,
            b'[' => TokenKind::LBracket,
            b']' => TokenKind::RBracket,
            b';' => TokenKind::Semi,
            b',' => TokenKind::Comma,
            0 => TokenKind::EOF,
            _ => TokenKind::UnexpectedCharacter,
        };
        self.bump();
        kind
    }

    fn ident(&mut self) -> TokenKind {
        self.bump_while(|chr| matches!(chr, b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_'));
        self.bump_while(|chr| chr == b'\'');

        let value = self.slice();

        match value {
            "and" => TokenKind::And,
            "or" => TokenKind::Or,
            "not" => TokenKind::Not,
            _ => {
                if value == "_" {
                    TokenKind::Underscore
                } else {
                    TokenKind::Identifier
                }
            }
        }
    }

    fn is_symbol(&self, chr: u8) -> bool {
        matches!(
            chr,
            b'=' | b':' | b'.' | b'|' | b'<' | b'>' | b'!' | b'+' | b'-' | b'*' | b'/' | b'^' | b'%'
        )
    }

    fn operator(&mut self) -> TokenKind {
        let kind = match self.curr {
            b'.' if self.peek == b'.' => {
                self.bump();
                TokenKind::Dots
            }
            b'.' => TokenKind::Dot,
            b'=' if self.peek == b'=' => {
                self.bump();
                TokenKind::EqEq
            }
            b'=' if self.peek == b'>' => {
                self.bump();
                TokenKind::Arrow
            }
            b'=' => TokenKind::Equals,
            b':' => TokenKind::Colon,
            b'<' if self.peek == b'=' => {
                self.bump();
                TokenKind::LtEq
            }
            b'<' => TokenKind::Lt,
            b'>' if self.peek == b'=' => {
                self.bump();
                TokenKind::GtEq
            }
            b'>' => TokenKind::Gt,
            b'!' if self.peek == b'=' => {
                self.bump();
                TokenKind::NoEq
            }
            b'+' => TokenKind::Plus,
            b'-' => TokenKind::Minus,
            b'*' => TokenKind::Star,
            b'/' => TokenKind::Slash,
            b'^' => TokenKind::Caret,
            b'%' => TokenKind::Mod,
            _ => unreachable!(),
        };

        self.bump();
        kind
    }

    fn number(&mut self) -> TokenKind {
        self.bump_while(|chr| chr.is_ascii_digit());

        match self.curr {
            b'.' if self.peek.is_ascii_digit() => {
                self.bump();
                self.bump_while(|chr| chr.is_ascii_digit());

                return TokenKind::Number;
            }
            b'b' if self.slice() == "0" => {
                self.bump();
                self.bump_while(|chr| matches!(chr, b'0' | b'1'));
            }
            b'x' if self.slice() == "0" => {
                self.bump();
                self.bump_while(|chr| chr.is_ascii_hexdigit());
            }
            _ => {}
        }

        TokenKind::Integer
    }

    fn string(&mut self) -> TokenKind {
        loop {
            match self.bump() {
                b'\n' | 0 => {
                    return TokenKind::UnterminatedString;
                }
                b'"' => {
                    self.bump();
                    break;
                }
                b'\\' => match self.bump() {
                    b'\\' | b'n' | b'r' | b't' | b'"' | b'0' => {}
                    b'u' => todo!("validate escaped unicode"),
                    b'x' => todo!("validate escaped binary"),
                    _ => {
                        return TokenKind::UnexpectedCharacter;
                    }
                },
                _ => {}
            }
        }

        TokenKind::String
    }

    fn newline(&mut self) -> TokenKind {
        while self.curr == b'\n' {
            self.bump();
            self.trivia();
        }
        TokenKind::EOL
    }

    fn trivia(&mut self) {
        loop {
            if self.curr.is_ascii_whitespace() && self.curr != b'\n' {
                self.bump_while(|chr| chr != b'\n' && chr.is_ascii_whitespace());
            } else if self.curr == b'/' && self.peek == b'/' {
                self.bump_while(|chr| chr != b'\n' && chr != b'\0');
            } else {
                break;
            }
        }
    }
}

pub fn tokens(source: &Source) -> Vec<Token> {
    let mut lexer = Lexer::new(source);
    let mut tokens = vec![];

    while let kind = lexer.kind() && kind != TokenKind::EOF {
        tokens.push(Token {
            kind,
            span: lexer.span(),
        });
    }

    tokens
}
